packages/dds/tree/docs/wip/Schema.ts

```diff
@@ -0,0 +1,74 @@
+/**
+ * Example internal schema representation types.
+ */
+
+/**
+ * Stable identifier, used when persisting data.
+ *
+ * This intentionally does not support any concept of versioning or name-spacing:
+ * users of it can include versions and namespaces in if they want, or features for them could be added later.
+ *
+ * Can be used either:
+ * 1. Record just this in the persisted data.
+ * When loading the data, see if the loader has a schema with a matching name, and if so, use that.
+ * Optionally validate loaded data against schema.
+ * 2. Persist the whole schema. Use the identifier to associate it with schema when loading to check that the schema match.
+ */
+export type SchemaIdentifier = string;
+
+export enum Multiplicity {
+    /**
+     * Exactly one item.
+     */
+    Value,
+    /**
+     * 0 or 1 items.
+     */
+    Optional,
+    /**
+     * 0 or more items.
+     */
+    Sequence,
+    /**
+     * Exactly 0 items.
+     */
+    Forbidden,
+}
+
+export enum Value {
+    Nothing,
+    Number,
+    String,
+    Boolean,
+    /**
+     * Any Fluid serializable data.
+     */
+    Serializable,
+}
+
+export interface FieldContent {
+    multiplicity: Multiplicity;
+    /**
+     * If not specified, child types are unconstrained.
+     */
+    types?: Type[];
+}
+
+export interface Field {
+    content: FieldContent;
+    /**
+     * Stable identifier, used when persisting data.
+     *
+     * Can be used either:
+     * 1. Record just this in the persisted data. When loading the data, see if the loader has a schema with a matching name, and if so, use that.
+     * 2. Persist the whole schema. Use the identifier to associate it with schema when loading to check that the schema match.
+     */
+    name: string;
+}
+
+export interface Type {
+    name: string;
+    fields: Field[];
+    extraFields?: FieldContent;
+    value: Value;
```

Comment:
Is `Value` a bit field, or will union types only be allowed within a field?

Reply:
No, it is not a bit field in this example code (but that would be a reasonable approach). This example is intentionally very minimal about what is supported functionality wise for values since how we want to handle them is still undecided (and this example isn't trying to take an opinion on that, just show an example that works): we expect them to only be used for custom primitives/terminals and not be a dimension application do anything with directly, so this makes them also likely to be an implementation detail.

Anyway, `ValueSchema.Serializable` is effectively 'any' for values, and since values do not support mutation (only replacing),  it should be straightforward for applications to have complex rules for values in specific types without needing to complicate the DDS with it (merges can't mutate values in unexpected ways like they can nodes, so values will stay in schema), so an app could just use  `ValueSchema.Serializable` and apply their own more restrictive rules in their `view schema`.